File: src/template/cpp/segtree/segtree.cpp
#include <vector>
using namespace std;

struct SegTree {
    int n;
    vector<long long> t;
    SegTree(int _n): n(_n), t(2*_n, 0) {}
    void update(int p, long long v) {
        p += n;
        t[p] += v;
        for (p >>= 1; p; p >>= 1)
            t[p] = t[p<<1] + t[p<<1|1];
    }
    long long query(int l, int r) const {
        long long res = 0;
        for (l += n, r += n; l < r; l >>= 1, r >>= 1) {
            if (l & 1) res += t[l++];
            if (r & 1) res += t[--r];
        }
        return res;
    }
};

File: src/template/julia/seg_tree/seg_tree.jl
struct SegTree
    n::Int
    t::Vector{Int}
end

function SegTree(n::Int)
    SegTree(n, zeros(Int, 2n))
end

function update!(st::SegTree, p::Int, v::Int)
    p += st.n
    st.t[p] += v
    while p > 1
        p ÷= 2
        st.t[p] = st.t[2p] + st.t[2p+1]
    end
end

function query(st::SegTree, l::Int, r::Int)
    res = 0
    l += st.n; r += st.n
    while l < r
        if isodd(l)
            res += st.t[l]; l += 1
        end
        if isodd(r)
            r -= 1; res += st.t[r]
        end
        l ÷= 2; r ÷= 2
    end
    return res
end